docs/proposals/nvidia-gpu-enhancement.md

```diff
@@ -0,0 +1,72 @@
+# Add NVIDIA GPU discovery and assignment capabilities to Kubelet.
+
+## Background:
+
+So far we only support 1 GPU(```/dev/nvidia0```) and the GPU device need to be configured as a Kubelet's parameter manually. But in production environment, there could have multiple GPUs(```/dev/nvidia0, /dev/nvidia1, /dev/nvidia2 ...```) installed on 1 Minion, also 1 micro service could needs more than 1 GPU, and there could have thousands of Minions which means if Kubelet could discover GPU by itself would be great.
+
+
+## Goal:
+
+ - Have the capabilities of GPU discovery to let Kubelet could aware more than 1 NVIDIA GPU and make the deployment easier.
+ - Assign more than 1 GPU to 1 container to make the micro service with multiple GPUs requirement run on Kubernetes possible.
+ - Add a general GPU interface for all the GPU vendors for future use.
+ - Add device support interface in runtime, so it could support both Docker and RKT.
+
+
+## Implementation:
+
+### Limitation:
+
+In this implementation, we still only schedule dedicated NVIDIA GPU and NVIDIA GPU supported only, and will not be NVIDIA GPU details(familiies, cores, memory) sensitive.
+
+### Features:
+
+We need to add below features to enhance GPU(especial NVIDIA GPU) capabilities. The discovery and assignment need to be tracked by separate modules in Kubelet.
+
+ 1. Discover all the NVIDIA GPUs on a Minion, but do not need all the NVIDIA GPU details information(families, cores, memory). To discover all the NVIDIA GPUs on a node, we can simply identify all devices on the host matching /dev/nvidiaX. By directly mapping through entire devices, we can avoid using the NVIDIA Management Library (NVML). In the future, we will likely want to add this library in order to monitor per-core utilization and take advantage of other features that will allow GPUs to be shared between containers.
+ 
+	1.1 Kubelet needs a list of GPU information, also need to report the number of free and total GPUs. Each general GPU device could be like:
+    ```go
+    type GPUDevice struct {
+		 ID		string	// GPU ID.
+	   Path string		 // The GPU path, like /dev/nvidia0, /dev/nvidia1, /dev/nvidia2...
+    }
+    ```
+
+	1.2 Once Kubelet restart, the GPU information needs to be recovered from containers' imformation.
+
+ 2. Assign more than 1 NVIDIA GPU to 1 container, and each container could have multiple NVIDIA GPUs. The devices host/container mapping also need to be added to the runtime information, just as the volume did. The container GPU mapping list could show the GPU assignment.
+	
```

Comment:
Clarify: a GPU can belong to at most one container.  If a pod has two containers, only one of them can have a GPU assigned to it.